### PR TITLE
Do not rate game or persist results for single player games

### LINKED ISSUE #175: Don't rate single player games
```
Rating groups: [{Player(Downlord, 146315, (1000.0, 100.0), (1000.0, 100.0)): trueskill.Rating(mu=1000.000, sigma=100.000)}]
Ranks: [0]
Error during game end: Need multiple rating groups
Traceback (most recent call last):
  File "/home/michel/IdeaProjects/server/server/games/game.py", line 283, in on_game_end
    await self.rate_game()
  File "/home/michel/IdeaProjects/server/server/games/custom_game.py", line 17, in rate_game
    new_ratings = self.compute_rating(rating='global')
  File "/home/michel/IdeaProjects/server/server/games/game.py", line 637, in compute_rating
    return trueskill.rate(rating_groups, ranks)
  File "/home/michel/py3.5/lib/python3.5/site-packages/trueskill/__init__.py", line 700, in rate
    return global_env().rate(rating_groups, ranks, weights, min_delta)
  File "/home/michel/py3.5/lib/python3.5/site-packages/trueskill/__init__.py", line 477, in rate
    rating_groups, keys = self.validate_rating_groups(rating_groups)
  File "/home/michel/py3.5/lib/python3.5/site-packages/trueskill/__init__.py", line 266, in validate_rating_groups
    raise ValueError('Need multiple rating groups')
ValueError: Need multiple rating groups
```

### PR DESCRIPTION
**Pull request by micheljung**
Fixes #175

### CODE REVIEW

File: server/games/game.py [hunk 1]
```diff
@@ -75,6 +75,7 @@ class ValidityState(IntEnum):
     BAD_MOD = 10
     COOP_NOT_RANKED = 11
     MUTUAL_DRAW = 12
+    SINGLE_PLAYER = 13
```
Comment (ChrisKitching) [inline server/games/game.py]: You never used this.

File: server/games/game.py [hunk 2]
```diff
@@ -271,13 +272,15 @@ def add_game_connection(self, game_connection):
...
 
                 if time.time() - self.launched_at > 4*60 and self.is_mutually_agreed_draw:
                     self._logger.info("Game is a mutual draw")
                     await self.mark_invalid(ValidityState.MUTUAL_DRAW)
 
-                await self.persist_results()
-                await self.rate_game()
+                if len(self.players) > 1:
+                    await self.persist_results()
+                    await self.rate_game()
```
Comment (ChrisKitching) [inline server/games/game.py]: You should probably invert this condition and call mark_invalid with `ValidityState.SINGLE_PLAYER`.

File: server/games/game.py [hunk 3]
```diff
@@ -271,13 +272,15 @@ def add_game_connection(self, game_connection):
 
                 if self.desyncs > 20:
                     await self.mark_invalid(ValidityState.TOO_MANY_DESYNCS)
+                    return
```
Comment (ChrisKitching) [inline server/games/game.py]: Due to this bug, you probably want to write a DB migration script to delete records from `game_player_stats` that refer to invalid games.

File: server/games/game.py [hunk 4]
```diff
@@ -271,13 +272,15 @@ def add_game_connection(self, game_connection):
...
 
                 if time.time() - self.launched_at > 4*60 and self.is_mutually_agreed_draw:
                     self._logger.info("Game is a mutual draw")
                     await self.mark_invalid(ValidityState.MUTUAL_DRAW)
```
Comment (ChrisKitching) [inline server/games/game.py]: This also needs a return.

File: server/games/game.py [hunk 5]
```diff
@@ -271,13 +272,15 @@ def add_game_connection(self, game_connection):
 
                 if self.desyncs > 20:
                     await self.mark_invalid(ValidityState.TOO_MANY_DESYNCS)
+                    return
 
                 if time.time() - self.launched_at > 4*60 and self.is_mutually_agreed_draw:
                     self._logger.info("Game is a mutual draw")
```
Comment (ChrisKitching) [inline server/games/game.py]: Isn't this pointless logspam?

File: server/games/game.py (hunk 2 above)
Comment (micheljung) [inline server/games/game.py]: That's actually what I wanted to do x) Thx

File: server/games/game.py (hunk 3 above)
Comment (micheljung) [inline server/games/game.py]: https://github.com/FAForever/db/issues/26

File: server/games/game.py (hunk 5 above)
Comment (micheljung) [inline server/games/game.py]: Yes. Like many others. At least most have been fixed to be `debug` instead of `info`

**Comment by Sheeo** [comment]
Is this good to merge @ChrisKitching?

**Comment by ChrisKitching** [comment]
> Is this good to merge @ChrisKitching?

No: an amendment hasn't yet been pushed.

**Comment by micheljung** [comment]
It's pushed now

**Comment by micheljung** [comment]
@Sheeo can we merge this?

**Comment by Sheeo** [comment]
@micheljung yes, we can.

**Pull request merged**